### ISSUE TITLE
Feature Request: Streamline image insert

### ISSUE THREAD

**Issue by memeticks** [jaap-karssenberg/zim-desktop-wiki#320]
I have recently discovered Zim and am currently using it to centralise what eventually will be a large amount of diverse data such as "how to" docs, config files, feature descriptions, release notes and various other useful tidbits. My typical usage is when I find a doc, spreadsheet, webpage etc that I need to be easy to find quickly I copy the contents and paste them into a Zim page. So far this is working beautifully for text. But when I want to insert images into Zim pages I discovered that when I drag/ drop images then Zim only links to the image and so when the image is deleted or moved it is lost from the Zim page. And also if I copy/ paste an image with some text the image doesn't get inserted at all. 
So when I want to insert an image from a doc or webpage I first have to copy the image, insert it into mspaint, save it, then go to attach file in Zim, and insert the file, and then finally delete the file. As you can imagine once this goes beyond 2 or 3 images then this becomes very time consuming and laborious and suddenly Zim is not going to work for me.
So my feature request is to make image inserting as easy as drag and drop, or copy paste. I have tested with word and libre office docs and if I copy text and images and then insert into a doc the image gets inserted along with the text as one would expect. So I am pretty sure the required data is there in the clipboard.
Otherwise Zim is perfect for my needs, and it would be a real shame if this stops me from using it.
Thanks for your time.

**Comment by Osndok** [comment]
To me, if *ever* zim creates a link to a file outside of the repo it is a bug. I hate, and am surprised whenever I find a broken image link in zim b/c of issues like this, or b/c the 'attach' checkbox was accidentally toggled, etc... and by the time you notice the issue the file is long gone!

**Comment by memeticks** [comment]
@Osndok - agreed, I cant think of any use case where I would prefer a link rather then a copy of the file, except maybe in years past when storage was expensive and devs had to be cognizant of storage requirements, but nowadays storage is so cheap its almost free, we no longer need to worry about a few MB or even GB here or there. I dont know if I use computers different to others but I am always moving folders/ files around in a constant struggle with managing data as my needs, goals, focus etc. change over time. The chances of links remaining current is slim to none for me, especially when talking about multi year time frames.
I have been testing many different offline notes apps and I think I have found one that handles my use case described above a little better http://keepnote.org/, it hasn't been updated in quite a while (2012) but so far it seems to work very well, it allows pasting of text with pictures and even preserves the text formatting such as bullets, fonts etc. I'm using the beta version 0.7.9 which seems to work better than 0.7.8. It also has a nifty screen capture feature which is great for stupid apps like google docs which wont allow copying of images to the clipboard. 
I guess Zim is designed for devs who want to just preserve code, in which case it works beautifully.

**Comment by Self-Perfection** [comment]
I disagree. "Just make a copy" creates clutter and chaos due to many versions of a file that might become unsynced. After inserting image to zim I might crop original image, enhance or somehow change it, edit metadata (geo coordinates, author, etc). That should be reflected in image that Zim points to.

Comparison to LibreOffice is not perfectly accurate. LO works with self contained documents, but Zim keeps its data structure transparent to OS in file/directory structure, I expect zim to integrate with directory tree of my system.

Though it might be a good idea to try to create hardlink and if hardlink creation fails give user a choice.

**Comment by Osndok** [comment]
git will dedupe a copy, move, copy-and-delete-original; and (to me) your argument about "chaos" actually favors keeping all zim file links local (with a page's directory), as that is the more truly more organized.

Hardlinks are a definite no-go... view few people would expect to edit file A and have file B changed, and they interfere with git anyway.

**Comment by jaap-karssenberg** [comment]
Probably this is one of those things that should be an option that users can choose their preferred behavior. Needs more concrete proposal under what conditions copies should be made and under what conditions links would be inserted when the option is enabled / disabled.

**Comment by jhe-** [comment]
Is the original issue still present? I'm using 0.68-rc1 and the behavior is pretty good in most cases:

- Browser (tested on Chrome):
  - Copy and paste: Stores a copy :heavy_check_mark: (alas, "Paste" is grayed out :x:)
  - Drag and drop: Inserts only the image URL (if the image is a link, the target URL is used instead) :grey_question:
- Image editor or screenshot:
  - Copy and paste: Stores a copy :heavy_check_mark: (alas, "Paste" is grayed out :x:)
  - Drag and drop: Stores a copy :heavy_check_mark:
- Locally stored image (This is the one where there are good arguments for both sides):
  - Copy and paste: Links to image path :grey_question:
  - Drag and drop: Links to image path :grey_question:
- Insert -> Image:
  - When "Attach image first" is unchecked: Links to image path :heavy_check_mark:
  - When "Attach image first" is checked: Stores a copy :heavy_check_mark:

(An unrelated nit, when the cursor is inline with an image, it becomes as large as the image itself even though the font size hasn't changed:

![image](https://user-images.githubusercontent.com/4308710/54854015-c551be80-4cae-11e9-9f4d-42c89e415537.png))

**Comment by kewball** [comment]
_Zim version 0.70_

When I choose the menu item _Insert/Image..._ the dialog box has the option "Attach image first"

I would like a preferences selection that I could set to make "Attach image first" the default action for drag'n'drop or copy/paste insertion of images. 

Unlike @jhe's experience above, none of those methods produces an attachment for me.

**Comment by jhe-** [comment]
@kewball If you're talking about image files stored on the local hard drive, then my experience is consistent. I only get attachments for image selections or web images, since there would be no path to refer to.

**Comment by kewball** [comment]
I mis-spoke. The Insert -> Image, click the checkbox for "attach image first" _does_ perform as advertised. It is quite verbose. Copy/paste or drag/drop would be much cleaner for my use case, but only if the image is attached.

I'd like the checkbox to be true by default for my notebooks.

**Comment by Korodny** [comment]
There are situations where Zim *has* to attach the image - i.e. whenever you're pasting an image from another application that is not a filemanager,

The other scenarios, where Zim could opt for either attaching or linking an image, are these:

- Insert->Image
- drag'n drop
- pasting from a filemanager

Currently, only the first approach is configurable, via the "Attach image" checkbox in the dialog box. This should be moved to the preferences dialog and affect all three of these scenarios, IMHO.

I think the default behaviour for Insert->Image changed recently without warning (images were no longer attached if you didn't select the checkbox once), that should be avoided because it might result in data loss. So, if this is made configurable, the default behaviour should be not to attach images.

**Comment by jaap-karssenberg** [comment]
Make option more obvious, also consider different actions for "copy & insert image" vs "link & insert image"

**Comment by jaap-karssenberg** [comment]
Also add option for behavior on drag&drop to attach or link by default

**Comment by casperl** [comment]
Instead of opening a separate issue, I believe my problem falls under this open issue, though perhaps not addressed in detail.

Right-Click menu - Paste greyed out despite image being in the clibpboard content

Problem:

1) Clipboard contains a copied image.
2) Attempting to paste the image into a Notebook page.
3) Right click over the Notebook page.
4) Paste is greyed out over in the popup menu.
5) Copy and paste works as expected for text both inside Zim and copying from another application and pasting into Zim.  Paste is only greyed out for images.

Workaround:

Manually pressing Control+V using the keyboard works. 

Reasoning:

1) Copy and paste to clipboard is a standard function in all operating systems.
2) It is fair that users will expect copy and paste to work consistently in applications.
3) Copying multiple images into Zim from an alternate app presently requires altering between the mouse and keyboard which becomes tedious when it could have been an operation accomplished by the mouse only.

Suggested Enhancement:

Also copy images into the attachment pane via the right click button.  This means a file is created and a folder/directory is created if it does not exist, if an image on the clipboard is pasted into the attachment pane by right-clicking in the attachment pane.

**Comment by hsandt** [comment]
Note that this related issue https://github.com/zim-desktop-wiki/zim-desktop-wiki/issues/884 is asking to actually properly show an image from URL (e.g. after drag-and-drop). So eventually we'll need to pick one behaviour for drag-and-drop:
a. Insert image by reference (show image via URL, not stored on disk except maybe a temporary copy)
b. Copy image content, storing asset on disk in page attachments

I found both options useful. There could be a prompt on first drag-and-drop to ask for a. or b., with a checkbox "Always apply this option (and don't ask anymore)" which could be changed later in the Preferences. Graphics editor sometimes prompt like that, sometimes they just link by default and you can copy content later in the tools menu (but it wouldn't work in Zim's more simple interface).
